qp_klp/Assays.py

```diff
@@ -0,0 +1,511 @@
+                        'data_type': None,
+                        'job-id': self.master_qiita_job_id,
+                        'name': afact_name}
+
+                if 'target_gene' in metadata[list(metadata.keys())[0]]:
+                    tg = metadata[list(metadata.keys())[0]]['target_gene']
+                    for key in Amplicon.AMPLICON_SUB_TYPES:
+                        if key in tg:
+                            data['data_type'] = key
+
+                    if data['data_type'] is None:
+                        raise ValueError("data_type could not be "
+                                         "determined from target_gene "
+                                         "column")
+                else:
+                    raise ValueError("target_gene must be specified for "
+                                     "amplicon type")
+
+                reply = self.qclient.post('/qiita_db/prep_template/',
+                                          data=data)
+                prep_id = reply['prep']
+                results[study_id].append((prep_id, afact_name, is_repl))
+                self.run_prefixes[prep_id] = [metadata[sample]['run_prefix']
+                                              for sample in metadata]
+
+        self.touched_studies_prep_info = results
+        return results
+
+    def load_preps_into_qiita(self):
+        data = []
+        for project, _, qiita_id in self.special_map:
+            fastq_files = self._get_postqc_fastq_files(
+                self.pipeline.output_path, project)
+
+            for vals in self.touched_studies_prep_info[qiita_id]:
+                prep_id, artifact_name, is_repl = vals
+                if is_repl:
+                    # for Amplicon runs, each prep needs a copy of the
+                    # entire set of fastq files, because demuxing samples
+                    # happens downstream. If we don't make copies of the
+                    # files, Qiita will move the files when loading the
+                    # first prep and they won't be available for the
+                    # second prep and after.
+                    # Note that this will leave the original files present
+                    # in the working directory after processing instead of
+                    # being moved.
+                    working_set = self._copy_files(fastq_files)
+                else:
+                    working_set = fastq_files
+
+                data.append(self._load_prep_into_qiita(
+                    self.qclient, prep_id, artifact_name, qiita_id, project,
+                    working_set, ARTIFACT_TYPE_AMPLICON))
+
+        df = pd.DataFrame(data)
+        opath = join(self.pipeline.output_path, 'touched_studies.html')
+        with open(opath, 'w') as f:
+            f.write(df.to_html(border=2, index=False, justify="left",
+                               render_links=True, escape=False))
+
+        return df
+
+
+class MetaOmic(Assay):
```

Comment:
I'm not 100% why we need Assay and MetaOmic; could you clarify?

Reply:
Sure. Historically what we've had are two (Amplicon, Metagenomic) then three (Metatranscriptomic) pipelines that are largely similar with but with several large and small differences along the path. One example is Amplicon skipping the host-filtering step and then having to copy and reorganize files so that the output can be fed into FastQC and load preps into Qiita downstream. Metagenomic and Metatranscriptomic are essentially identical but they are required to have individual types. Now TellSeq is essentially the same pipeline downstream of host-filtering but the conversion process is entirely different and the host filtering needed to be modified to support its metadata.

Since we like to stay 'dry' and not repeat ourselves in code, we can't define four different pipelines that are 80% identical. Previous versions of SPP resolved this with an ever growing number of 'if type == 'Amplicon' else:' type statements. However introducing new features potentially broke all pathways and it became harder to understand how to modify the code to add new features or fix bugs.

What Assays does is break out all the code that changes depending on the pipeline being Amplicon or Meta*omic and organizes it so that it's all easier to read, changes to one type won't affect the other so it doesn't need to be retested, and we don't need to constantly test whether we're an amplicon run or not. Moreover helper methods that are shared between the different pipelines are defined in the base Assay class and don't have to be duplicated in each class.